### ISSUE TITLE
support file url

### ISSUE THREAD

**Issue by dnalborczyk** [rollup/rollup#4278]
### Feature Use Case

was just playing with: https://github.com/rollup/rollup/issues/4275 and noticed file URLs are not supported.
 
I think in the _short therm_ rollup could implement support by converting the file url to a path, similar to https://github.com/rollup/rollup/issues/4275#issuecomment-972549237

I think in the long run it might be beneficial to use file URLs as cache key, then other protocols (e.g. http) could be potentially supported on top as well.

```js
// index.js
import foo from "file:///foo/bar/index.js";

export default {};

console.log(foo, import.meta.url);
```

```shell
node index.js

{} file:///foo/bar/index.js
```

```js
// rollup.config.js
export default {
  input: "./index.js",
  output: {
    file: "./build/bundle.js",
    name: "bundle",
  },
}
```

```js
// bundle.js
import foo from 'file:///foo/bar/index.js';

const index = {};

console.log(foo, import.meta.url);

export { index as default };
```

```shell
node build/bundle.js

{} file:///foo/bar/index.js
{} file:///foo/bar/build/bundle.min.js
```

### Feature Proposal

.

**Comment by lukastaegert** [comment]
I assume "are not supported" means just the same as for queries and hashes: Rollup does not know how to read them from disk? I guess it makes sense to support them the same way as query parameters.

> I think in the long run it might be beneficial to use file URLs as cache key

About that I am not 100% sure. One could try out things here but my fear is that a lot of ad-hoc plugins will just break, making a lot of users unhappy if the ids passed to plugins are suddenly urls. And in the end, for most users it is not important how Rollup resolves imports as it is actually the node-resolve plugin that creates most of the ids. So if one would ever want to go there, this could be a way forward:

* audit all "official" plugins so that they work nicely with file urls as well as file paths
* make a breaking node-resolve version that changes things and see how much people hate us :) at that point, people can just not upgrade or make their stuff compatible, or we can revert the change
* If enough time passed (> half a year AT LEAST, rather more), thing about making the change in core

**Comment by dnalborczyk** [comment]
> I assume "are not supported" means just the same as for queries and hashes: Rollup does not know how to read them from disk? I guess it makes sense to support them the same way as query parameters.

yeah, rollup would need to convert those file URLs to file paths and/or vice versa. (https://nodejs.org/dist/latest-v17.x/docs/api/url.html#urlfileurltopathurl https://nodejs.org/dist/latest-v17.x/docs/api/url.html#urlpathtofileurlpath . although some of the `fs` and `path` api does support URLs.

`node.js` also uses URLs for esm: https://nodejs.org/dist/latest-v17.x/docs/api/esm.html#urls, same as browsers.

> > I think in the long run it might be beneficial to use file URLs as cache key
> 
> About that I am not 100% sure. One could try out things here but my fear is that a lot of ad-hoc plugins will just break, making a lot of users unhappy if the ids passed to plugins are suddenly urls. And in the end, for most users it is not important how Rollup resolves imports as it is actually the node-resolve plugin that creates most of the ids. So if one would ever want to go there, this could be a way forward:
> 
> * audit all "official" plugins so that they work nicely with file urls as well as file paths

one potential avenue could also be to implement URL support internally for rollup first, and pass on the `file path` to the plugins as always. I have not looked into what the contract looks like between rollup and the plugins, but maybe it's possible to pass the URL in addition to the file path to the plugins?

> * make a breaking node-resolve version that changes things and see how much people hate us :) 

🤣 

at that point, people can just not upgrade or make their stuff compatible, or we can revert the change
> * If enough time passed (> half a year AT LEAST, rather more), thing about making the change in core

I think this should be probably be part of v3, in case stuff breaks. 

speaking of node-resolve: that reminds me that rollup is diverging as well from making the file extension _not_ mandatory. even though the browser cares less about a file extension, but it would need it to resolve the file location nonetheless (unless the server does an extension rewrite). I'm almost thinking that the extension should be mandatory as well, and maybe have a fallback thru a flag or something, where the current order of file extension resolution can be specified [.js, .mjs] etc.

**Comment by lukastaegert** [comment]
> one potential avenue could also be to implement URL support internally for rollup first, and pass on the file path to the plugins as always. I have not looked into what the contract looks like between rollup and the plugins, but maybe it's possible to pass the URL in addition to the file path to the plugins?

I do not know what this would mean. You cannot use them as ids "just internally" because these need to be the same ids that plugins use, otherwise you start breaking some core concepts of Rollup's ids. If a plugin explicitly uses urls as ids, and there are people who do that already, they should expect to receive urls as ids. But support for resolving and loading them when they are used explicitly sounds great to me and will be the first step anyway.

> speaking of node-resolve: that reminds me that rollup is diverging as well from making the file extension not mandatory

That is kind of a traditional thing. I do not see people liking it very much when we enforce that now, though. Especially since TypeScript more or less trained people to NOT use file extensions by penalizing them when they do. But this has been a long-standing pain point.
